geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/utils/Reprojection.scala

```diff
@@ -16,43 +16,90 @@ import org.locationtech.geomesa.features.ScalaSimpleFeature
 import org.locationtech.geomesa.utils.geotools.SimpleFeatureTypes
 import org.locationtech.jts.geom.Geometry
 import org.opengis.feature.simple.{SimpleFeature, SimpleFeatureType}
+import org.opengis.referencing.crs.CoordinateReferenceSystem
 
-class Reprojection private (sft: SimpleFeatureType, transformer: Option[GeometryCoordinateSequenceTransformer]) {
-  def reproject(feature: SimpleFeature): SimpleFeature = {
-    val values = Array.tabulate(sft.getAttributeCount) { i =>
-      feature.getAttribute(i) match {
-        case g: Geometry => transformer.map(_.transform(g)).getOrElse(g)
-        case a => a
-      }
-    }
-    val result = ScalaSimpleFeature.create(sft, feature.getID, values: _*)
-    result.getUserData.putAll(feature.getUserData)
-    result
-  }
+/**
+  * Reproject the geometries in a simple feature to a different CRS
+  */
+trait Reprojection {
+  def apply(feature: SimpleFeature): SimpleFeature
 }
 
 object Reprojection {
 
   import org.locationtech.geomesa.index.conf.QueryHints.RichHints
 
-  def apply(query: Query): Option[Reprojection] = {
-    val sft = query.getHints.getReturnSft
-    if (sft.getGeometryDescriptor == null) { None } else {
-      val native = sft.getGeometryDescriptor.getCoordinateReferenceSystem
-      val source = Option(query.getCoordinateSystem).getOrElse(native)
-      val target = Option(query.getCoordinateSystemReproject).getOrElse(native)
-
-      if (target != source) {
-        val transformer = new GeometryCoordinateSequenceTransformer
-        transformer.setMathTransform(CRS.findMathTransform(source, target, true))
-        val reprojected = SimpleFeatureTypes.immutable(FeatureTypes.transform(sft, target), sft.getUserData)
-        Some(new Reprojection(reprojected, Some(transformer)))
-      } else if (source != native) {
-        val reprojected = SimpleFeatureTypes.immutable(FeatureTypes.transform(sft, source), sft.getUserData)
-        Some(new Reprojection(reprojected, None))
-      } else {
-        None
+  /**
+    * Create a reprojection function
+    *
+    * @param returnSft simple feature type being returned
+    * @param crs crs information from a query
+    * @return
+    */
+  def apply(returnSft: SimpleFeatureType, crs: QueryReferenceSystems): Reprojection = {
+    if (crs.target != crs.user) {
+      val transformer = new GeometryCoordinateSequenceTransformer
+      transformer.setMathTransform(CRS.findMathTransform(crs.user, crs.target, true))
+      val transformed = FeatureTypes.transform(returnSft, crs.target) // note: drops user data
```

Comment:
Which userData is dropped?

Reply:
it doesn't copy the sft user data, hence we add it back in the next line.